### PR TITLE
Implements clamp

### PR DESCRIPTION
**Pull request by redmoogle**
Hopefully I did this right

made using regretful web commits

### CODE REVIEW

File: OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs [hunk 1]
```diff
@@ -79,6 +80,17 @@ static class DreamProcNativeRoot {
 
             return new DreamValue(Convert.ToChar(ascii).ToString());
         }
+        
+        [DreamProc("clamp")]
+        [DreamProcParameter("Value", Type = DreamValueType.Number)]
+        [DreamProcParameter("Low", Type = DreamValueType.Number)]
+        [DreamProcParameter("High", Type = DreamValueType.Number)]
+        public static DreamValue NativeProc_ckey(DreamObject instance, DreamObject usr, DreamProcArguments arguments) {
```
Comment (MarkSuckerberg) [inline OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs:88]:
```suggestion
        public static DreamValue NativeProc_clamp(DreamObject instance, DreamObject usr, DreamProcArguments arguments) {
```

File: DMCompiler/DMStandard/_Standard.dm [hunk 2]
```diff
@@ -5,6 +5,7 @@ proc/animate(Object, time, loop, easing, flags)
 proc/arccos(X)
 proc/arctan(A)
 proc/ascii2text(N)
+proc/clamp(Value, Low, High)
```
Comment (wixoaGit) [inline DMCompiler/DMStandard/_Standard.dm:8]: To keep this in alphabetic order, move this under `proc/ckey()`

File: OpenDreamServer/Dream/Procs/Native/DreamProcNative.cs [hunk 3]
```diff
@@ -10,6 +10,7 @@ static class DreamProcNative {
             root.SetNativeProc(DreamProcNativeRoot.NativeProc_arccos);
             root.SetNativeProc(DreamProcNativeRoot.NativeProc_arctan);
             root.SetNativeProc(DreamProcNativeRoot.NativeProc_ascii2text);
+            root.SetNativeProc(DreamProcNativeRoot.NativeProc_clamp);
```
Comment (wixoaGit) [inline OpenDreamServer/Dream/Procs/Native/DreamProcNative.cs:13]: Also move this below ckey

File: OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs (hunk 1 above)
Comment (redmoogle) [reply OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs:88]: Am fixing already

Comment (wixoaGit) [inline OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs:88]: `clamp()` can also take a list as its first argument. It needs to handle this.

Also, move this below ckey.

File: OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs [hunk 4]
```diff
@@ -6,6 +6,7 @@
 using System;
 using System.Collections.Generic;
 using System.Collections.Specialized;
+using System.Math;
```
Comment (wixoaGit) [inline OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs:9]: Math is a class in the System namespace, this isn't valid.

File: OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs [hunk 5]
```diff
@@ -79,6 +80,17 @@ static class DreamProcNativeRoot {
...
+        
+        [DreamProc("clamp")]
+        [DreamProcParameter("Value", Type = DreamValueType.Number)]
+        [DreamProcParameter("Low", Type = DreamValueType.Number)]
+        [DreamProcParameter("High", Type = DreamValueType.Number)]
+        public static DreamValue NativeProc_ckey(DreamObject instance, DreamObject usr, DreamProcArguments arguments) {
+            DreamValue iValue = arguments.GetArgument(0, "Value");
+            DreamValue lowValue = arguments.GetArgument(1, "Low");
+            DreamValue highValue = arguments.GetArgument(2, "High");
+            return Math.Clamp(iValue, lowValue, highValue);
```
Comment (wixoaGit) [inline OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs:92]: Arguments are DreamValues, and have to be unwrapped. Similarly, a proc's return value must be wrapped in a DreamValue.

```cs
float iValue = arguments.GetArgument(0, "Value").GetValueAsNumber();
float lowValue = arguments.GetArgument(1, "Low").GetValueAsNumber();
float highValue = arguments.GetArgument(2, "High").GetValueAsNumber();
return new DreamValue(Math.Clamp(iValue, lowValue, highValue));
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: TestGame/environment.dme [hunk 6]
```diff
@@ -1,3 +1,12 @@
...
+#include "interface.dmf"
+// BEGIN_INTERNALS
+// END_INTERNALS
+// BEGIN_FILE_DIR
+#define FILE_DIR .
+// END_FILE_DIR
+// BEGIN_PREFERENCES
+// END_PREFERENCES
+// BEGIN_INCLUDE
+// END_INCLUDE
```
Comment (ZeWaka) [inline TestGame/environment.dme:12]: stop using DM aaaaaaaaaaaa

File: OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs [hunk 7]
```diff
@@ -79,6 +79,31 @@ static class DreamProcNativeRoot {
 
             return new DreamValue(Convert.ToChar(ascii).ToString());
         }
+        
+        [DreamProc("clamp")]
+        [DreamProcParameter("Value")]
```
Comment (wixoaGit) [inline OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs:84]: This parameter can be either a number, or a DreamList (a DreamObject)

```suggestion
        [DreamProcParameter("Value", Type = DreamValueType.Number | DreamValueType.DreamObject)]
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs [hunk 8]
```diff
@@ -89,6 +89,32 @@ static class DreamProcNativeRoot {
...
+            float hVal = arguments.GetArgument(2, "High").GetValueAsNumber();
+
+            if (value.TryGetValueAsDreamList(out DreamList list))
+            {
+                DreamList tmp = new DreamList();
+                for (int i = 1; i < list.GetLength(); i++)
+                {
+                    DreamValue val = list.GetValues()[i];
+                    tmp.AddValue(new DreamValue((float)Math.Clamp(val.GetValueAsNumber(), lVal, hVal)));
+                }
```
Comment (wixoaGit) [inline OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs:109]:
```suggestion
                foreach (DreamValue val in list.GetValues()) {
                    tmp.AddValue(new DreamValue(Math.Clamp(val.GetValueAsNumber(), lVal, hVal)));
                }
```

File: OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs [hunk 9]
```diff
@@ -89,6 +89,32 @@ static class DreamProcNativeRoot {
...
+                return new DreamValue(tmp);
+            }
+            else
+            {
+                return new DreamValue((float)Math.Clamp(value.GetValueAsNumber(), lVal, hVal));
```
Comment (wixoaGit) [inline OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs:114]: `Math.Clamp` already returns a float if all three arguments are floats.
```suggestion
                return new DreamValue(Math.Clamp(value.GetValueAsNumber(), lVal, hVal));
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**